lib/go-parser/go-binary.ts

```diff
@@ -0,0 +1,401 @@
+import * as depGraph from "@snyk/dep-graph";
+import { eventLoopSpinner } from "event-loop-spinner";
+import * as path from "path";
+import * as varint from "varint";
+
+import { DEP_GRAPH_TYPE } from "./";
+import { GoModule } from "./go-module";
+import { LineTable } from "./pclntab";
+import { Elf, ElfProgram } from "./types";
+
+export class GoBinary {
+  public name: string;
+  public modules: GoModule[];
+
+  constructor(goElfBinary: Elf) {
+    [this.name, this.modules] = extractModuleInformation(goElfBinary);
+
+    const pclnTab = goElfBinary.body.sections.find(
+      (section) => section.name === ".gopclntab",
+    );
+    if (pclnTab === undefined) {
+      throw Error("no pcln table found in Go binary");
+    }
+
+    this.matchFilesToModules(new LineTable(pclnTab.data).go12MapFiles());
+  }
+
+  public async depGraph(): Promise<depGraph.DepGraph> {
+    const goModulesDepGraph = new depGraph.DepGraphBuilder(
+      { name: DEP_GRAPH_TYPE },
+      { name: this.name },
+    );
+
+    for (const module of this.modules) {
+      for (const pkg of module.packages) {
+        if (eventLoopSpinner.isStarving()) {
+          await eventLoopSpinner.spin();
+        }
+
+        const version = module.snykNormalisedVersion();
+        const nodeId = `${pkg}@${version}`;
+        goModulesDepGraph.addPkgNode({ name: pkg, version }, nodeId);
+        goModulesDepGraph.connectDep(goModulesDepGraph.rootNodeId, nodeId);
+      }
+    }
+
+    return goModulesDepGraph.build();
+  }
+
+  // matchFilesToModules goes through all files, extracts the package name and
+  // adds it to the relevant module in the GoBinary.
+  public matchFilesToModules(files: string[]): void {
+    // goModCachePath is the path at which the modules are downloaded to. When
+    // building a Go binary, this is usually either $GOMODCACHE or
+    // $GOROOT/pkg/mod. Binaries built with `-trimpath` will have that module
+    // cache path trimmed away, meaning that the goModCachePath will always be
+    // empty.
+    let goModCachePath: string = "";
+    let moduleName = (mod: GoModule): string => {
```

Comment:
It is because we need to use this function on all modules that we have. The module name is usually `my.modules.com/module@v1.0.0`, but in vendored binaries it is simply `my.modules.com/module`. 

This is for *all* modules in a binary, which is why we need to use a different function to get that name.